File: BFS/Populating Next Right Pointers in Each Node/Populating Next Right Pointers in Each Node.java

/*
// Definition for a Node.
class Node {
    public int val;
    public Node left;
    public Node right;
    public Node next;

    public Node() {}
    
    public Node(int _val) {
        val = _val;
    }

    public Node(int _val, Node _left, Node _right, Node _next) {
        val = _val;
        left = _left;
        right = _right;
        next = _next;
    }
};
*/
import java.util.*;
class Solution {
    public Node connect(Node root) {
        if(root ==null) return root;
        Queue<Node> q = new LinkedList<>(); 
        q.offer(root);
        while(!q.isEmpty()){
            int n = q.size();
            Node prev = null;
            for(int i =0; i<n; i++){
                Node temp = q.peek();
                if(q.peek().left != null)
                    q.offer(q.peek().left);
                if(q.peek().right != null)
                    q.offer(q.peek().right);
                q.poll();
                if( prev != null) prev.next = temp;
                prev = temp;
            }
        }
        return root;
    }
}

File: BFS/Populating Next Right Pointers in Each Node/Populating Next Right Pointers in Each Node.cpp

/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* left;
    Node* right;
    Node* next;

    Node() : val(0), left(NULL), right(NULL), next(NULL) {}

    Node(int _val) : val(_val), left(NULL), right(NULL), next(NULL) {}

    Node(int _val, Node* _left, Node* _right, Node* _next)
        : val(_val), left(_left), right(_right), next(_next) {}
};
*/
class Solution {
public:
    Node* connect(Node* root) {
        queue<Node*> q;
        q.push(root);
        while(!q.empty()){
            int n = q.size();
            Node* prev = NULL;
            while(n--){
                Node* temp = q.front();
                q.pop();
                if(temp->left) q.push(temp->left);
                if(temp->right) q.push(temp->right);
                if(prev) prev->next = temp;
                prev = temp;
                
            }
        }
        return root;
    }
};
